In [15]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [16]:
# Load the given datasets
ckd_cat = pd.read_csv('data/chronic_kidney_disease_categorical.csv')
ckd_num = pd.read_csv('data/chronic_kidney_disease_numerical.csv')
# Print the data
print(ckd_cat)
print(ckd_num)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
Each data seems to represent a person that had various medical readings taken, such as blood pressure, sugar level, blood cell counts, etc. The features are these measurements, with there being 24 of them total. The label would be the "class", which is whether the individual hadchronic kidney disease or not. Some of these features are numerical and others are categorical. Three examples of the categorical features are specific gravity (sg), albumin (al), and sugar (su). Three examples of numerical features include age, blood pressure (bp), and blood glucose random (bgr).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [17]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_dup_series = ckd_num.duplicated(subset=['unique_id'], keep=False)

print(f'Number of duplicate rows for numeric data set: {num_dup_series.sum()}')

# Drop these duplicate rows

ckd_num_no_dups = ckd_num.drop_duplicates(subset='unique_id')


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_dup_series = ckd_cat.duplicated(subset='unique_id', keep=False)

print(f'Number of duplicate rows for categorical data set: {cat_dup_series.sum()}')
print(ckd_cat[cat_dup_series]) # Printing the duplicate rows

ckd_cat_no_dups = ckd_cat.drop_duplicates(subset='unique_id')



Number of duplicate rows for numeric data set: 59
Number of duplicate rows for categorical data set: 59
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
10      708255  2.0  0.0    normal    normal  notpresent  notpresent   no   
34      441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
39      240975  0.0  0.0    normal    normal  notpresent  notpresent   no   
51      354223  0.0  0.0    normal    normal  notpresent  notpresent   no   
71      283427  0.0  0.0    normal    normal  notpresent  notpresent   no   
74      302906  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
88      856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
94      501321  1.0  0.0       NaN    normal  notpresent  notpresent   no   
113     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
124     261137  0.0  0.0    normal    normal  not

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
No, not all of the unique IDs are present in both data sets. We know this because one of the datasets has 400 rows, while the other only has 390, so there must be more unique IDs in one than the other. When one of the unique IDs is only present in one of the data sets, the columns from the other data set are filled with NaN, indicating that there was no value for that particular column with that unique_id.

In [18]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

merged_data = pd.merge(left=ckd_num_no_dups, right=ckd_cat_no_dups, on='unique_id', how='outer')

#Print the combined dataset
print(merged_data)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [19]:
# Calculate the percentage of rows that contain atleast one missing value
rows_with_na = merged_data.isna().any(axis=1)
num_rows = rows_with_na.sum()

percentage = num_rows / len(merged_data) * 100


# Print %

print(f'Percentage of rows with missing value: {percentage}')

# Drop these rows from the dataset

merged_no_na = merged_data.dropna(axis=0, how='any')

# Print the Dataset
print(merged_no_na)


Percentage of rows with missing value: 61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ..

## Part 3.4: Sort the dataset according to the Labels

In [20]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

merged_no_na.sort_values(by='Target', inplace=True, ignore_index=True)

# Print the dataset

print(merged_no_na)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

/tmp/ipykernel_17357/3363519319.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_no_na.sort_values(by='Target', inplace=True, ignore_index=True)


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [21]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
merged_dummies = pd.get_dummies(merged_no_na, columns=categorical_columns, dtype='int')

# Print the dataset
print(merged_dummies)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.


For binary variables, I think having the values seperated into multiple columns, as we have done, would lead to similar results. But by having them separated, they would act as new features that the regression would have to find appropriate values of theta for. If they were in a single column, there would only be a single theta, and in that way the modelling would be different. Similarly, with five values, that would lead to five more "features" that have to be modeled for during the regression, where if they were in a single column, it would only act as a single feature and a single theta value.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [22]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

interested_columns = merged_dummies.filter(items=numerical_columns, axis=1)
means = interested_columns.mean(axis=0)
deviations = interested_columns.std(axis=0)

merged_no_outliers = merged_dummies

for col_name in numerical_columns:
    upper_limit = means[col_name] + (3 * deviations[col_name])
    lower_limit = means[col_name] - (3 * deviations[col_name])
    
    merged_no_outliers = merged_no_outliers[(merged_no_outliers[col_name] > lower_limit) & (merged_no_outliers[col_name] < upper_limit)]

merged_no_outliers.reset_index(inplace=True, drop=True)

# Print the dataset

print(merged_no_outliers)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
1       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
2       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
3       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
4       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
129     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
130     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
131     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
132     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
133     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [23]:
# Normalize the all Numerical Attributes in the dataset.

for col_name in numerical_columns:
    cur_series = merged_no_outliers[col_name]

    max_val = cur_series.max()
    min_val = cur_series.min()

    merged_no_outliers[col_name] = cur_series.apply(lambda x: (x - min_val) / (max_val - min_val))

# Print the dataset
print(merged_no_outliers)

     unique_id       age    bp       bgr        bu        sc       sod  \
0       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
129     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
130     203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
131     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
132     228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
133     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
0    0.657143  0.172131  0.210526  ...  

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:
The unique_id column is unnecessary, since it doesn't provide any actual information on any factors that could impact CKD, and is instead metadeta for the table. The rest are all potential factors that could impact someone's health, and should therefore be left in the table.

In [24]:
#Remove that column
final_data = merged_no_outliers.drop(labels=['unique_id'], axis=1)

# Print the dataset
print(final_data)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
1    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.171429   
2    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  0.828571   
3    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
4    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.485714   
..        ...   ...       ...       ...       ...       ...       ...   
129  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.257143   
130  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.400000   
131  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.371429   
132  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.171429   
133  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.514286   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
0    0.172131  0.210526  0.395161  ...       0      

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [25]:
# Export the dataframe to a new csv file
final_data.to_csv('out_data_csv.csv', index=False)
# Export the dataframe to a new json file
final_data.to_json('out_data_json.json', index=False)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here
| | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|----------|------|----------|--------|------------|-----------|-----------|-------------|------------|-------------|-------------|--------|-------|----------------|---------------|----------------|-------------|------------------|----------------|----------------|--------------|----------|-----------|---------|----------|----------|-----------|---------------|---------------|---------|----------|----------|-----------|-----------|-----------------|
| 0 | 0.743243 | 0.5 | 0.442060 | 0.902 | 0.432099 | 0.5 | 0.657143 | 0.172131 | 0.210526 | 0.395161 | 0.153846 | 2.0 | 0.0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 1 | 0.770270 | 1.0 | 0.901288 | 0.163 | 0.345679 | 0.767 | 0.171429 | 0.606557 | 0.631579 | 0.443548 | 0.410256 | 2.0 | 2.0 | 0 | 1 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 2 | 0.905405 | 0.5 | 0.785408 | 0.863 | 0.518519 | 0.6 | 0.828571 | 0.401639 | 0.447368 | 0.233871 | 0.435897 | 2.0 | 0.0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 3 | 0.202703 | 0.75| 0.158798 | 0.196 | 0.160494 | 0.167 | 0.171429 | 0.221311 | 0.184211 | 0.653226 | 0.333333 | 4.0 | 0.0 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 |
| 4 | 0.783784 | 1.0 | 0.399142 | 0.288 | 0.839506 | 0.667 | 0.485714 | 0.188525 | 0.263158 | 0.258065 | 0.205128 | 4.0 | 2.0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 |
| 5 | 0.729730 | 0.0 | 0.935622 | 0.170 | 0.160494 | 0.333 | 0.028571 | 0.188525 | 0.236842 | 0.879032 | 0.102564 | 3.0 | 1.0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 |
| 6 | 0.675676 | 0.75| 0.253219 | 0.634 | 0.777778 | 0.367 | 0.542857 | 0.286885 | 0.342105 | 0.169355 | 0.205128 | 2.0 | 0.0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 |
| 7 | 0.851351 | 0.25| 0.618026 | 0.562 | 0.728395 | 0.0 | 0.285714 | 0.311475 | 0.315789 | 0.580645 | 0.179487 | 4.0 | 3.0 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 1 |
| 8 | 0.540541 | 0.0 | 0.399142 | 0.536 | 0.358025 | 0.7 | 0.314286 | 0.344262 | 0.315789 | 0.830645 | 0.153846 | 1.0 | 0.0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 |
| 9 | 0.756757 | 0.25| 0.223176 | 0.209 | 0.160494 | 0.533 | 0.514286 | 0.573770 | 0.605263 | 0.290323 | 0.333333 | 3.0 | 0.0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 |
| 10 | 0.662162 | 0.5 | 0.618026 | 0.412 | 0.432099 | 0.567 | 0.571429 | 0.434426 | 0.473684 | 0.25 | 0.282051 | 3.0 | 1.0 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 11 | 0.783784 | 0.0 | 0.725322 | 0.314 | 0.481481 | 0.567 | 0.714286 | 0.319672 | 0.342105 | 0.258065 | 0.205128 | 4.0 | 1.0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 |
| 12 | 0.878378 | 0.0 | 0.206009 | 0.752 | 0.604938 | 0.533 | 0.571429 | 0.475410 | 0.5 | 0.879032 | 0.435897 | 4.0 | 0.0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 13 | 0.878378 | 0.25| 0.639485 | 0.471 | 0.395062 | 0.433 | 0.428571 | 0.393443 | 0.447368 | 0.104839 | 0.256410 | 3.0 | 0.0 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 14 | 0.716216 | 0.5 | 1.0 | 0.163 | 0.111111 | 0.067 | 0.171429 | 0.393443 | 0.5 | 0.532258 | 0.435897 | 1.0 | 0.0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

One way I can think about mitigating this is by having the language model regenerate an output for the same data multiple times, and then based on the differences between the outputs you could more easily spot hallucinations, since they would be inconsistent. Another possible mitigation technique could be by breaking large data sets into smaller pieces, since its more likely to hallucinate the larger the data set is.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [26]:
# Code to rename all the columns in the dataset

final_data.rename(columns={
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',  # Note that 'sg' is in your mapping, but it's not in the given data
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',   # Also note that 'rbc' is mentioned but doesn't appear in the given data.
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',  # This should be 'wc' based on the mapping, but is 'wbcc' in your data.
    'rbcc': 'red blood cell count',  # This should be 'rc' based on the mapping, but is 'rbcc' in your data.
    'htn_no': 'hypertension no',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
}, inplace=True)

print(final_data)

          age  blood pressure  blood glucose random  blood urea  \
0    0.743243            0.50              0.442060    0.901961   
1    0.770270            1.00              0.901288    0.163399   
2    0.905405            0.50              0.785408    0.862745   
3    0.202703            0.75              0.158798    0.196078   
4    0.783784            1.00              0.399142    0.287582   
..        ...             ...                   ...         ...   
129  0.189189            0.25              0.227468    0.222222   
130  0.216216            0.00              0.115880    0.052288   
131  0.513514            0.00              0.107296    0.235294   
132  0.702703            0.50              0.128755    0.261438   
133  0.689189            0.00              0.150215    0.254902   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
0            0.432099  0.500000   0.657143    0.172131            0.210526   
1            0.345679  0.766667   0.171

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [27]:
# Converted SQL to Pandas code
# Assuming final_data is your existing DataFrame
target_counts = ckd_cat.groupby('Target').size().reset_index(name='count')

# Displaying the result
print(target_counts)

   Target  count
0     ckd    264
1   ckd\t      2
2  notckd    164


The above code seems to count the two values of Target, ckd and notckd (and also ckd\t), and returns a new data_frame that holds these counts.